File: Behavioral/Lab22/C#/TemplateMethods/AddApartmentTemplate.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Entities;

namespace C_.TemplateMethods
{
    public class AddApartmentTemplate : AbstractTemplate
    {
        public AddApartmentTemplate(params string[] args) : base(args){}


        protected override string AddEmpty()
        {
            var agency = Agency.GetInstance();
            agency.AddEstate(new Apartment());
            return "Added empty new Apartment";
        }

        protected override string AddEstate()
        {
            if(_params.Length != 6) 
                return ReturnError();
            var agency = Agency.GetInstance();
            agency.AddEstate(new Apartment(_params[0], _params[1], Convert.ToDouble(_params[2]), Convert.ToDouble(_params[3]), _params[4], Convert.ToInt16(_params[5])));
            return "Added new Apartment with args: " + string.Join(", ", _params);
        }

        protected override string ReturnError(){
            return "Failed to add new Apartment with args: " + string.Join(", ", _params);
        }
    }
}

File: Behavioral/Lab22/Typescript/src/TemplateMethods/add_apartment_template.ts
import { Agency } from "../Entities/agency";
import { Apartment } from "../Entities/apartment";
import { AbstractTemplate } from "./abstract_template";

export class AddApartmentTemplate extends AbstractTemplate {
    constructor(...params: string[]) {
        super(...params);
    }

    protected override addEmpty(): string {
        const agency = Agency.GetInstance();
        agency.AddEstate(new Apartment());
        return "Added empty new Apartment";
    }

    protected override addEstate(): string {
        const agency = Agency.GetInstance();
        if (this.Params.length === 6) {
            agency.AddEstate(new Apartment(
                this.Params[0],
                this.Params[1],
                +this.Params[2],
                +this.Params[3],
                this.Params[4],
                +this.Params[5]
            ));
            return "Added new Apartment with args: " + this.Params.join(", ");
        }
        else
            return this.returnError();
    }

    protected override returnError(): string {
        return "Failed to add new Apartment with args: " + this.Params.join(", ");
    }
}
